pkg/cache/sql/db/client.go

```diff
@@ -0,0 +1,321 @@
+		rows, err = stmt.QueryContext(ctx, params...)
+		if err != nil {
+			sqlErr, ok := err.(*sqlite.Error)
+			if ok && sqlErr.Code() == sqlite3.SQLITE_BUSY {
+				return false, nil
+			}
+			return false, errors.Wrapf(err, "Error initializing Store DB")
+		}
+		return true, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return rows, nil
+}
+
+// CloseStmt will call close on the given Closable. It is intended to be used with a sql statement. This function is meant
+// to replace stmt.Close which can cause panics when callers unit-test since there usually is no real underlying connection.
+func (c *Client) CloseStmt(closable Closable) error {
+	return closable.Close()
+}
+
+// ReadObjects Scans the given rows, performs any necessary decryption, converts the data to objects of the given type,
+// and returns a slice of those objects.
+func (c *Client) ReadObjects(rows Rows, typ reflect.Type, shouldDecrypt bool) ([]any, error) {
+	c.connLock.RLock()
+	defer c.connLock.RUnlock()
+	// defer rows.Close()
+
+	// rt := time.Now()
+	var result []any
+	for rows.Next() {
+		data, err := c.decryptScan(rows, shouldDecrypt)
+		if err != nil {
+			return nil, closeRowsOnError(rows, err)
+		}
+		singleResult, err := fromBytes(data, typ)
+		if err != nil {
+			return nil, closeRowsOnError(rows, err)
+		}
+		result = append(result, singleResult.Elem().Interface())
+	}
+	err := rows.Err()
+	if err != nil {
+		return nil, closeRowsOnError(rows, err)
+	}
+
+	err = rows.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// ReadStrings scans the given rows into strings, and then returns the strings as a slice.
+func (c *Client) ReadStrings(rows Rows) ([]string, error) {
+	c.connLock.RLock()
+	defer c.connLock.RUnlock()
+
+	var result []string
+	for rows.Next() {
+		var key string
+		err := rows.Scan(&key)
+		if err != nil {
+			ce := rows.Close()
+			if ce != nil {
+				return nil, errors.Wrap(ce, "while handling "+err.Error())
+			}
+			return nil, err
+		}
+
+		result = append(result, key)
+	}
+	err := rows.Err()
+	if err != nil {
+		ce := rows.Close()
+		if ce != nil {
+			return nil, errors.Wrap(ce, "while handling "+err.Error())
+		}
+		return nil, err
+	}
+
+	err = rows.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// Begin attempt to begin a transaction, and returns it along with a function for unlocking the
+// database once the transaction is done.
+func (c *Client) Begin() (TXClient, error) {
+	c.connLock.RLock()
+	defer c.connLock.RUnlock()
+	sqlTx, err := c.conn.Begin()
+	if err != nil {
+		return nil, err
+	}
+	return transaction.NewClient(sqlTx), nil
+}
+
+func (c *Client) decryptScan(rows Rows, shouldDecrypt bool) ([]byte, error) {
+	var data, dataNonce, eDEK, dekNonce sql.RawBytes
+	err := rows.Scan(&data, &dataNonce, &eDEK, &dekNonce)
+	if err != nil {
+		return nil, err
+	}
+	if c.decryptor != nil && shouldDecrypt {
+		data, err := c.decryptor.Decrypt(data, dataNonce, eDEK, dekNonce)
```

Comment:
I'm not a fan of shadowed non-error variables. Could you make this `data` something like `decryptedData` ?